[WPE] Add explicit fence support for wayland

alexgcastro opened the pull request:
#### 863d0f1f6ab1bcfdbc47b4a689af051c337576ad
<pre>
[WPE] Add explicit fence support for wayland
<a href="https://bugs.webkit.org/show_bug.cgi?id=276818">https://bugs.webkit.org/show_bug.cgi?id=276818</a>

Reviewed by Miguel Gomez.

We are adding support to pass fence from the WebProcess to the wayland
server to avoid stopping the pipeline in the CPU. This is the initial
implementation, we are using the protocol
linux-explicit-synchronization-unstable-v1 and we plan to add support
for new protocols and DRM fences in the future.

The compositor generates a fence for the frame that is used to create a
synchronization point in the command queue and passed to the wayland
server. Also we receive the fence for the buffer usage from the wayland
server that we associate to the buffer and pused to the command queue
before using the buffer.

The patch adds the parameters to the messages in the IPC communication
between WebProcess and UIProces and methods to handle both fences.

Co-authored-by: Carlos Garcia Campos &lt;cgarcia@igalia.com&gt;

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/dmabuf/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::useExplicitSync const):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
(WebKit::AcceleratedBackingStoreDMABuf::bufferReleased):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::useExplicitSync const):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
(wpe_buffer_dma_buf_set_rendering_fence):
(wpe_buffer_dma_buf_get_rendering_fence):
(wpe_buffer_dma_buf_take_rendering_fence):
(wpe_buffer_dma_buf_set_release_fence):
(wpe_buffer_dma_buf_get_release_fence):
(wpe_buffer_dma_buf_take_release_fence):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_use_explicit_sync):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandUseExplicitSync):
(wpeDisplayWaylandGetLinuxExplicitSync):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
(wpeToplevelWaylandConstructed):
(wpeToplevelWaylandDispose):
(wpeToplevelWaylandHasSyncObject):
(wpeToplevelWaylandGetSyncListener):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(DMABufBuffer::DMABufBuffer):
(DMABufBuffer::~DMABufBuffer):
(dmaBufBufferDestroy):
(createWaylandBufferFromDMABuf):
(SharedMemoryBuffer::SharedMemoryBuffer):
(sharedMemoryBufferCreate):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::useExplicitSync const):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::setReleaseFence):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::releaseTarget):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::releaseBuffer):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.messages.in:

Canonical link: <a href="https://commits.webkit.org/281640@main">https://commits.webkit.org/281640@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56454007168bcd0c93f050bee3293bae50cda766

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60343 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64263 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10875 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11108 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48857 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7574 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62374 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36998 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52279 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29704 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33696 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9792 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9793 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65995 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4277 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9630 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56216 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4295 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56384 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3564 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9102 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35505 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36586 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37676 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36330 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -630,8 +630,11 @@ bool AcceleratedBackingStoreDMABuf::prepareForRendering()
+            auto finishRenderingFence = UnixFileDescriptor { };
+            m_webPage.legacyMainFrameProcess().send(Messages::AcceleratedSurfaceDMABuf::ReleaseBuffer(m_committedBuffer->id(), WTFMove(finishRenderingFence)), m_surfaceID);
+        }
+
```
```suggestion
        if (m_committedBuffer)          m_webPage.legacyMainFrameProcess().send(Messages::AcceleratedSurfaceDMABuf::ReleaseBuffer(m_committedBuffer->id(), { }), m_surfaceID);

```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
```diff
@@ -168,8 +170,13 @@ void AcceleratedBackingStoreDMABuf::bufferRendered()
-    if (auto id = m_bufferIDs.get(buffer))
-        m_webPage.legacyMainFrameProcess().send(Messages::AcceleratedSurfaceDMABuf::ReleaseBuffer(id), m_surfaceID);
+    if (auto id = m_bufferIDs.get(buffer)) {
+        UnixFileDescriptor finishRenderingFence;
```
I would call this releaseFence

carlosgcampos commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -192,6 +192,18 @@ OptionSet<WebCore::PlatformEvent::Modifier> WebPageProxy::currentStateOfModifier
+    auto* view = wpeView();
+    if (!view)
+        return false;
+    return wpe_display_use_explicit_sync(wpe_view_get_display(view));
```
```suggestion
    if (auto* view = wpeView())
        return wpe_display_use_explicit_sync(wpe_view_get_display(view));
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -192,6 +192,18 @@ OptionSet<WebCore::PlatformEvent::Modifier> WebPageProxy::currentStateOfModifier
+    return wpe_display_use_explicit_sync(wpe_view_get_display(view));
+#else
+    return false;
+#endif
```
With previous suggestion we don't need the `#else` just `#endif` and return false

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
```diff
@@ -53,6 +53,8 @@ struct _WPEBufferDMABufPrivate {
     Vector<uint32_t> strides;
     uint64_t modifier;
     EGLImage eglImage;
+    UnixFileDescriptor finishRenderingFence;
```
```suggestion
    UnixFileDescriptor renderingFence;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
```diff
@@ -382,3 +384,99 @@ guint64 wpe_buffer_dma_buf_get_modifier(WPEBufferDMABuf* buffer)
+ * @fd: a file descriptor
+ *
+ * Set the sync file descriptor to use for the @buffer.
+ * The buffer takes the ownership of the file descriptor.
```
I guess -1 is a valid value too, we should probably document it.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
```diff
@@ -857,6 +864,16 @@ void wpeToplevelWaylandViewVisibilityChanged(WPEToplevelWayland* toplevel, WPEVi
+struct zwp_linux_surface_synchronization_v1* wpeToplevelWaylandGetSyncListener(WPEToplevelWayland* toplevel)
+{
+    return toplevel->priv->sync;
+}
```
These could be just one, a getter of the sync object that can also be used to check if it's present.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -31,6 +31,7 @@
 #include "WPEToplevelWaylandPrivate.h"
 #include "WPEWaylandSHMPool.h"
 #include "linux-dmabuf-unstable-v1-client-protocol.h"
+#include "linux-explicit-synchronization-unstable-v1-client-protocol.h"
```
This is not needed here since it's included from private header already.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -156,57 +182,72 @@ static struct wl_buffer* createWaylandBufferFromEGLImage(WPEView* view, WPEBuffe
 
-    wl_buffer_add_listener(wlBuffer, &bufferListener, buffer);
+    auto* toplevel = wpe_view_get_toplevel(WPE_VIEW(view));
+    if ((toplevel && !wpeToplevelWaylandHasSyncObject(WPE_TOPLEVEL_WAYLAND(toplevel))) || wpe_buffer_dma_buf_get_rendering_fence(bufferDMABuf) == -1)
```
render buffer can't be called without a toplevel, so we don't need to null-check it here.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -298,6 +356,16 @@ static gboolean wpeViewWaylandRenderBuffer(WPEView* view, WPEBuffer* buffer, con
     auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
     wl_surface_attach(wlSurface, wlBuffer, 0, 0);
 
+    auto* syncListener = wpeToplevelWaylandGetSyncListener(WPE_TOPLEVEL_WAYLAND(wpe_view_get_toplevel(view)));
```
Is this a listener? I would call this `wlSync` or `wlSurfaceSync`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -298,6 +356,16 @@ static gboolean wpeViewWaylandRenderBuffer(WPEView* view, WPEBuffer* buffer, con
     wl_surface_attach(wlSurface, wlBuffer, 0, 0);
 
+    auto* syncListener = wpeToplevelWaylandGetSyncListener(WPE_TOPLEVEL_WAYLAND(wpe_view_get_toplevel(view)));
+    auto finishRenderingFence = UnixFileDescriptor { wpe_buffer_dma_buf_take_rendering_fence(WPE_BUFFER_DMA_BUF(buffer)), UnixFileDescriptor::Adopt };
```
```suggestion
    auto renderingFence = UnixFileDescriptor { wpe_buffer_dma_buf_take_rendering_fence(WPE_BUFFER_DMA_BUF(buffer)), UnixFileDescriptor::Adopt };
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -284,6 +322,26 @@ const struct wl_callback_listener frameListener = {
+        auto* buffer = WPE_BUFFER(userData);
+        wpe_view_buffer_released(wpe_buffer_get_view(buffer), buffer);
+        if (auto* dmaBufBuffer = static_cast<DMABufBuffer*>(wpe_buffer_get_user_data(buffer)))
+            g_clear_pointer(&dmaBufBuffer->release, zwp_linux_buffer_release_v1_destroy);
```
This is duplicated code we could move it to a helper and call it from both release callbacks. For example, `wpe_view_wayland_buffer_released`

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -595,6 +605,22 @@ void AcceleratedSurfaceDMABuf::willRenderFrame()
 
+    if (m_useExplicitSync && m_target->supportsExplicitSync()) {
+        auto fence = m_target->takeReleaseFence();
+        if (fence != -1) {
```
We should use UnixFileDescriptor here, either make takeReleaseFence return a UnixFileDescriptor or adopt here the returned fd in a new UnixFileDescriptor

carlosgcampos replied:
Another possibility is doing all the fence wait in the target adding `RenderTarget::waitRelease` or something like that

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -606,18 +632,26 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(const std::optional<WebCore::Regio
-    if (auto fence = WebCore::GLFence::create(WebCore::GLFence::ShouldFlush::No))
-        fence->wait(WebCore::GLFence::FlushCommands::Yes);
-    else
+    UnixFileDescriptor didRenderFrameFence;
```
```suggestion
    UnixFileDescriptor renderingFence;
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -606,18 +632,26 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(const std::optional<WebCore::Regio
+        glFlush();
+        if (fence != EGL_NO_SYNC_KHR) {
+            didRenderFrameFence = UnixFileDescriptor { eglDupNativeFenceFDANDROID(display.eglDisplay(), fence), UnixFileDescriptor::Adopt };
+            eglWaitSyncKHR(display.eglDisplay(), fence, 0);
```
Why do we need to wait here?

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -606,18 +632,26 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(const std::optional<WebCore::Regio
+            eglWaitSyncKHR(display.eglDisplay(), fence, 0);
+            eglDestroySyncKHR(display.eglDisplay(), fence);
+        }
+    } else
         glFlush();
```
When explicit sync is not available or not supported by the target or we fail to create the fence we should use a normal fence and wait like we did before. Again, maybe it would be easier to handle the fence in the target adding `RenderTarget::createRenderingFence`

alexgcastro committed:
[WPE] Add explicit fence support for wayland
https://bugs.webkit.org/show_bug.cgi?id=276818

Reviewed by Miguel Gomez.

We are adding support to pass fence from the WebProcess to the wayland
server to avoid stopping the pipeline in the CPU. This is the initial
implementation, we are using the protocol
linux-explicit-synchronization-unstable-v1 and we plan to add support
for new protocols and DRM fences in the future.

The compositor generates a fence for the frame that is used to create a
synchronization point in the command queue and passed to the wayland
server. Also we receive the fence for the buffer usage from the wayland
server that we associate to the buffer and pused to the command queue
before using the buffer.

The patch adds the parameters to the messages in the IPC communication
between WebProcess and UIProces and methods to handle both fences.

Co-authored-by: Carlos Garcia Campos <cgarcia@igalia.com>

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/dmabuf/AcceleratedBackingStoreDMABuf.messages.in:
* Source/WebKit/UIProcess/gtk/WebPageProxyGtk.cpp:
(WebKit::WebPageProxy::useExplicitSync const):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::frame):
(WebKit::AcceleratedBackingStoreDMABuf::bufferReleased):
* Source/WebKit/UIProcess/wpe/AcceleratedBackingStoreDMABuf.h:
* Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
(WebKit::WebPageProxy::useExplicitSync const):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.cpp:
(wpe_buffer_dma_buf_set_rendering_fence):
(wpe_buffer_dma_buf_get_rendering_fence):
(wpe_buffer_dma_buf_take_rendering_fence):
(wpe_buffer_dma_buf_set_release_fence):
(wpe_buffer_dma_buf_get_release_fence):
(wpe_buffer_dma_buf_take_release_fence):
* Source/WebKit/WPEPlatform/wpe/WPEBufferDMABuf.h:
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_use_explicit_sync):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandUseExplicitSync):
(wpeDisplayWaylandGetLinuxExplicitSync):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWayland.cpp:
(wpeToplevelWaylandConstructed):
(wpeToplevelWaylandDispose):
(wpeToplevelWaylandHasSyncObject):
(wpeToplevelWaylandGetSyncListener):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEToplevelWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(DMABufBuffer::DMABufBuffer):
(DMABufBuffer::~DMABufBuffer):
(dmaBufBufferDestroy):
(createWaylandBufferFromDMABuf):
(SharedMemoryBuffer::SharedMemoryBuffer):
(sharedMemoryBufferCreate):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::useExplicitSync const):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::AcceleratedSurfaceDMABuf):
(WebKit::AcceleratedSurfaceDMABuf::RenderTarget::setReleaseFence):
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::releaseTarget):
(WebKit::AcceleratedSurfaceDMABuf::willRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::releaseBuffer):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.messages.in:

Canonical link: https://commits.webkit.org/281640@main
